### PR TITLE
Use ingress

### PR DESCRIPTION
**Pull request by sed-i**
This PR:
- adds support for the ingress relation (traefik's ingress per **app**)
- adds a config option, `web_external_url`, to support any other ingress.

The updated data flow is depicted in the diagram below.

image

Crossref: [OPENG-11](https://warthogs.atlassian.net/browse/OPENG-11), [OPENG-28](https://warthogs.atlassian.net/browse/OPENG-28)

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -18,3 +18,20 @@ options:
...
+      prefix all HTTP endpoints served by Alertmanager.
+      If omitted, relevant URL components will be derived automatically.
+      If provided, this should be a complete URI, including scheme, or a
+      fully qualified subpath starting with `/`.
+      If Alertmanager is being served directly from the root of a fully-qualified
+      host or a bare A record, this may be omitted.
+      This configuration option takes precedence over the URL provided over
+      the "ingress" relation.
+    type: string
+    default: ""
```
Comment (jnsgruk) [inline config.yaml:37]: This shouldn't be included when you're using Traefik ingress - this is assigned automatically when related to Traefik, and can be customised by the (human) operator by using `traefik-route-k8s`.

Comment (sed-i) [reply config.yaml:37]: You mean drop the entire config option?
- Seems like [traefik-route](https://charmhub.io/traefik-route-k8s) is intended for ingress-per-unit, while here we use ingress-per-app.
- Seems like traefik-route needs to be deployed together with traefik. The idea behind the config option is to support any ingress (we won't be able to charm all ingresses in a timely manner).

File: requirements.txt [hunk 2]
```diff
@@ -1,7 +1,27 @@
...
+# Deps: charm
 ops
+
+# YAML processing framework
+# Code: https://github.com/yaml/pyyaml
+# Docs: https://pyyaml.org/wiki/PyYAMLDocumentation
+# Deps: charm
 PyYAML
+
+# Lightweight k8s module
```
Comment (balbirthomas) [inline requirements.txt:22]: Nice idea to add this comments.

File: tests/integration/test_templates.py [hunk 3]
```diff
@@ -93,34 +95,41 @@ async def test_receiver_gets_alert(ops_test: OpsTest, charm_under_test, httpserv
...
-            "/", method="POST", json=expected_notification
-        ).respond_with_handler(request_handler)
+        httpserver.expect_oneshot_request("/", method="POST").respond_with_handler(request_handler)
         client_address = await get_unit_address(ops_test, app_name, 0)
         amanager = Alertmanager(address=client_address)
         amanager.set_alerts(alerts)
 
     # check receiver got an alert
     assert waiting.result
+    assert request_from_alertmanager["attachments"][0]["callback_id"] == callback_id  # type: ignore
```
Comment (balbirthomas) [inline tests/integration/test_templates.py:135]: I presume you have checked that this test can be made to fail by changing the expected `callback_id` to be different from the `callback_id` that is set by the template.

**Review by balbirthomas — APPROVED** [review]

File: config.yaml (hunk 1 above)
Comment (sed-i) [reply config.yaml]: Updated description.

File: requirements.txt [hunk 4]
```diff
@@ -1,7 +1,27 @@
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
 
+# Operator Framework
+# Code: https://github.com/canonical/operator/
+# Docs: https://ops.rtfd.io/
+# Deps: traefik_k8s
+jsonschema
```
[the remark this answers is not in the document]
Comment (sed-i) [reply requirements.txt]: Fixed in next commit.

File: tests/integration/helpers.py [hunk 5]
```diff
@@ -95,3 +95,17 @@ async def is_alertmanager_up(ops_test: OpsTest, app_name: str):
...
+#     pebble_service_name = "alertmanager"
+#
+#     rc, stdout, stderr = await ops_test.juju(
+#         "exec", f"--unit={app_name}/{unit_num}", "--",
+#         f"PEBBLE_SOCKET=/charm/containers/{workload_container_name}/pebble.socket",
+#         "pebble", "plan",
+#     )
+#
+#     pebble_plan = yaml.safe_load(stdout)
+#     return pebble_plan["services"][pebble_service_name]
```
[the remark this answers is not in the document]
Comment (sed-i) [reply tests/integration/helpers.py]: Oops :)

**Pull request merged**